custom_components/unifiprotect/config_flow.py

```diff
@@ -40,51 +41,93 @@ def async_get_options_flow(config_entry):
 
+    async def async_step_reauth(self, user_input=None):
+        """Perform reauth upon an API authentication error."""
+        return await self._init_config_entry(user_input=user_input, reauth=True)
+
     async def async_step_user(self, user_input=None):
         """Handle a flow initiated by the user."""
-        if user_input is None:
-            return await self._show_setup_form(user_input)
+        return await self._init_config_entry(user_input=user_input)
 
-        errors = {}
+    def _get_form(self, reauth: bool):
+        form = self._show_setup_form
+        if reauth:
+            form = self._show_reauth_form
+
+        return form
 
+    async def _get_protect(self, user_input) -> Optional[UpvServer]:
         session = async_create_clientsession(
             self.hass, cookie_jar=CookieJar(unsafe=True)
         )
 
-        unifiprotect = UpvServer(
-            session,
-            user_input[CONF_HOST],
-            user_input[CONF_PORT],
-            user_input[CONF_USERNAME],
-            user_input[CONF_PASSWORD],
+        if self.unique_id is not None:
+            entry = await self.async_set_unique_id(self.unique_id)
+            host = entry.data[CONF_HOST]
+            port = entry.data[CONF_PORT]
+        elif CONF_HOST in user_input and CONF_PORT in user_input:
+            host = user_input[CONF_HOST]
+            port = user_input[CONF_PORT]
+        else:
+            return None
+
+        return UpvServer(
+            session=session,
+            host=host,
+            port=port,
+            username=user_input[CONF_USERNAME],
+            password=user_input[CONF_PASSWORD],
         )
 
+    async def _reload_entry(self, unique_id, user_input):
+        entry = await self.async_set_unique_id(unique_id)
+
+        new_data = entry.data.copy()
+        new_data[CONF_USERNAME] = user_input[CONF_USERNAME]
+        new_data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+
+        self.hass.config_entries.async_update_entry(entry, data=new_data)
+        await self.hass.config_entries.async_reload(entry.entry_id)
+        return self.async_abort(reason="reauth_successful")
+
+    async def _init_config_entry(self, user_input=None, reauth=False):
+        """Common method to initalize config entry"""
+
+        form = self._get_form(reauth)
+        if user_input is None:
+            return await form(user_input)
+
+        errors = {}
+        protect = await self._get_protect(user_input)
+
+        if protect is None:
+            return await form(errors)
+
         try:
-            server_info = await unifiprotect.server_information()
+            server_info = await protect.server_information()
             if server_info["server_version"] < MIN_REQUIRED_PROTECT_V:
                 _LOGGER.debug("UniFi Protect Version not supported")
                 errors["base"] = "protect_version"
-                return await self._show_setup_form(errors)
+                return await form(errors)
```

Comment:
Usually we build these so everything falls through and shows the form at the end if there are errors.

Simple example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/august/config_flow.py#L123

Reply:
Probably a better example: 
https://github.com/home-assistant/core/blob/dev/homeassistant/components/harmony/config_flow.py#L62